libraries/lib-utility/Composite.h

```diff
@@ -0,0 +1,66 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/**********************************************************************
+ 
+ Audacity: A Digital Audio Editor
+ 
+ @file Composite.h
+ @brief Support for the Composite pattern
+ 
+ Paul Licameli
+ 
+ **********************************************************************/
+#ifndef __AUDACITY_COMPOSITE__
+#define __AUDACITY_COMPOSITE__
+
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace Composite {
+
+//! Generates a base class for internal nodes of tree structures, which acts
+//! like a standard container (including compatibility with range-for and
+//! back_inserter)
+/*!
+ @tparam Component common base class of nodes and leaves (and of Base itself),
+ which must have a virtual destructor
+ @tparam Item type of pointer to Components to be stored
+ @tparam ComponentArgs... passed to constructor of Component
+ */
+template<
+   typename Component,
+   typename Item,
```

Comment:
So probably we can call it `ComponentPtr` or `PointerType`?

Reply:
Easily done, if you like, only affecting the template parameter locally.

Inside the class, it's aliased to `value_type`, named that specifically to make it work with `std::back_inserter`.